File: Day-18/order-saga-sample/python/order_saga.py
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum


class SagaStatus(Enum):
    COMPLETED = "COMPLETED"
    FAILED = "FAILED"


@dataclass(frozen=True)
class SagaResult:
    status: SagaStatus
    history: list[str]


class SagaStep:
    def __init__(self, name: str) -> None:
        self.name = name

    def execute(self) -> bool:
        raise NotImplementedError

    def compensate(self) -> bool:
        raise NotImplementedError


class OrderSagaOrchestrator:
    def __init__(self, steps: list[SagaStep]) -> None:
        if not steps:
            raise ValueError("steps must be non-empty")
        self._steps = steps[:]

    def run(self) -> SagaResult:
        completed: list[SagaStep] = []
        history: list[str] = []

        for step in self._steps:
            history.append(f"execute {step.name}")
            ok = step.execute()
            history.append(f"result {step.name}: {'OK' if ok else 'FAIL'}")
            if not ok:
                history.append(f"saga failed at {step.name}; starting compensation")
                for completed_step in reversed(completed):
                    history.append(f"compensate {completed_step.name}")
                    compensated = completed_step.compensate()
                    history.append(
                        f"compensation {completed_step.name}: {'OK' if compensated else 'FAIL'}"
                    )
                return SagaResult(status=SagaStatus.FAILED, history=history)
            completed.append(step)

        history.append("saga completed")
        return SagaResult(status=SagaStatus.COMPLETED, history=history)


class InventoryStep(SagaStep):
    def __init__(self) -> None:
        super().__init__("inventory")
        self._reserved = False

    def execute(self) -> bool:
        self._reserved = True
        return True

    def compensate(self) -> bool:
        if self._reserved:
            self._reserved = False
        return True


class PaymentStep(SagaStep):
    def __init__(self, should_fail: bool = False) -> None:
        super().__init__("payment")
        self._should_fail = should_fail
        self._charged = False

    def execute(self) -> bool:
        if self._should_fail:
            return False
        self._charged = True
        return True

    def compensate(self) -> bool:
        if self._charged:
            self._charged = False
        return True


class ShippingStep(SagaStep):
    def __init__(self, should_fail: bool = False) -> None:
        super().__init__("shipping")
        self._should_fail = should_fail
        self._created = False

    def execute(self) -> bool:
        if self._should_fail:
            return False
        self._created = True
        return True

    def compensate(self) -> bool:
        if self._created:
            self._created = False
        return True

File: Day-18/order-saga-sample/java/OrderSaga.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class OrderSaga {
    public enum SagaStatus {
        COMPLETED,
        FAILED
    }

    public static final class SagaResult {
        private final SagaStatus status;
        private final List<String> history;

        private SagaResult(SagaStatus status, List<String> history) {
            this.status = status;
            this.history = history;
        }

        public SagaStatus getStatus() {
            return status;
        }

        public List<String> getHistory() {
            return history;
        }
    }

    public interface SagaStep {
        String name();
        boolean execute();
        boolean compensate();
    }

    public static final class OrderSagaOrchestrator {
        private final List<SagaStep> steps;

        public OrderSagaOrchestrator(List<SagaStep> steps) {
            if (steps == null || steps.isEmpty()) {
                throw new IllegalArgumentException("steps must be non-empty");
            }
            this.steps = new ArrayList<>(steps);
        }

        public SagaResult run() {
            List<SagaStep> completed = new ArrayList<>();
            List<String> history = new ArrayList<>();

            for (SagaStep step : steps) {
                history.add("execute " + step.name());
                boolean ok = step.execute();
                history.add("result " + step.name() + ": " + (ok ? "OK" : "FAIL"));
                if (!ok) {
                    history.add("saga failed at " + step.name() + "; starting compensation");
                    Collections.reverse(completed);
                    for (SagaStep completedStep : completed) {
                        history.add("compensate " + completedStep.name());
                        boolean compensated = completedStep.compensate();
                        history.add(
                            "compensation "
                                + completedStep.name()
                                + ": "
                                + (compensated ? "OK" : "FAIL")
                        );
                    }
                    return new SagaResult(SagaStatus.FAILED, history);
                }
                completed.add(step);
            }

            history.add("saga completed");
            return new SagaResult(SagaStatus.COMPLETED, history);
        }
    }

    public static final class InventoryStep implements SagaStep {
        private boolean reserved = false;

        @Override
        public String name() {
            return "inventory";
        }

        @Override
        public boolean execute() {
            reserved = true;
            return true;
        }

        @Override
        public boolean compensate() {
            if (reserved) {
                reserved = false;
            }
            return true;
        }
    }

    public static final class PaymentStep implements SagaStep {
        private final boolean shouldFail;
        private boolean charged = false;

        public PaymentStep(boolean shouldFail) {
            this.shouldFail = shouldFail;
        }

        @Override
        public String name() {
            return "payment";
        }

        @Override
        public boolean execute() {
            if (shouldFail) {
                return false;
            }
            charged = true;
            return true;
        }

        @Override
        public boolean compensate() {
            if (charged) {
                charged = false;
            }
            return true;
        }
    }

    public static final class ShippingStep implements SagaStep {
        private final boolean shouldFail;
        private boolean created = false;

        public ShippingStep(boolean shouldFail) {
            this.shouldFail = shouldFail;
        }

        @Override
        public String name() {
            return "shipping";
        }

        @Override
        public boolean execute() {
            if (shouldFail) {
                return false;
            }
            created = true;
            return true;
        }

        @Override
        public boolean compensate() {
            if (created) {
                created = false;
            }
            return true;
        }
    }
}
